### PR TITLE
Interface phase2 - implement managed mode

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

**Comment by aramase** [comment]
```
Ran 19 of 20 Specs in 3549.162 seconds
SUCCESS! -- 19 Passed | 0 Failed | 0 Pending | 1 Skipped
--- PASS: TestAADPodIdentity (3549.16s)
PASS
ok	github.com/Azure/aad-pod-identity/test/e2e	3550.169s
```

File: cmd/nmi/main.go [hunk 1]
```diff
@@ -101,3 +107,20 @@ func main() {
 		klog.Fatalf("%s", err)
 	}
 }
+
+func getTokenClient(client k8s.Client) nmi.TokenClient {
+	klog.Infof("Initializing in %s mode", strings.ToLower(*operationMode))
```
Comment (kkmsft) [inline cmd/nmi/main.go]: Can we please perform a ToLower once at the beginning as soon as we get the parameters ?

File: pkg/nmi/server/managed.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		}
+	}
+	return nil, fmt.Errorf("no matching azure identity found for pod")
+}
+
+// GetToken ...
+func (mc *ManagedClient) GetToken(ctx context.Context, rqClientID, rqResource string, podID aadpodid.AzureIdentity) (token *adal.Token, err error) {
+	rqHasClientID := len(rqClientID) != 0
+	clientID := podID.Spec.ClientID
+	if rqHasClientID && !strings.EqualFold(rqClientID, clientID) {
```
Comment (kkmsft) [inline pkg/nmi/server/managed.go]: Would this check for fail for Service Principal all the time since it would not have clientID ?

Comment (kkmsft) [reply pkg/nmi/server/managed.go]: or it might pass because both are empty, but not a required check for that code path ?

File: pkg/nmi/nmi.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+const (
+	// StandardMode ...
+	StandardMode OperationMode = "standard"
+	// ManagedMode ...
+	ManagedMode OperationMode = "managed"
+)
+
+// TokenClient ...
+type TokenClient interface {
+	// GetIdentities gets the list of identities which match the
```
Comment (kkmsft) [inline pkg/nmi/nmi.go]: Can we please add a README and link it to the features list which explains the token interface and the split. Some short text there would be great to help newer developers get more clarity on what is going on.

File: pkg/nmi/server/standard.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	return nil, true, fmt.Errorf("getting assigned identities for pod %s/%s in ASSIGNED state failed after %d attempts, retry duration [%d]s. Error: %v",
+		podns, podname, sc.ListPodIDsRetryAttemptsForCreated+sc.ListPodIDsRetryAttemptsForAssigned, sc.ListPodIDsRetryIntervalInSeconds, err)
+}
+
+// GetToken ...
+func (sc *StandardClient) GetToken(ctx context.Context, rqClientID, rqResource string, podID aadpodid.AzureIdentity) (token *adal.Token, err error) {
+	rqHasClientID := len(rqClientID) != 0
+	clientID := podID.Spec.ClientID
+	if rqHasClientID && !strings.EqualFold(rqClientID, clientID) {
```
Comment (kkmsft) [inline pkg/nmi/server/standard.go]: Can we please de-dupe the code here with the code in managed mode ?

**Comment by kkmsft** [comment]
Are we planning to have separate PR for helm chart changes ? (would really prefer that :-) )

File: go.mod [hunk 5]
```diff
@@ -14,6 +14,7 @@ require (
 	github.com/coreos/go-iptables v0.3.0
 	github.com/gogo/protobuf v1.2.1 // indirect
 	github.com/golang/groupcache v0.0.0-20180513044358-24b0969c4cb7 // indirect
+	github.com/google/go-cmp v0.3.0
 	github.com/google/gofuzz v0.0.0-20170612174753-24818f796faf // indirect
 	github.com/googleapis/gnostic v0.1.0 // indirect
 	github.com/howeyc/gopass v0.0.0-20170109162249-bf9dde6d0d2c // indirect
```
[the remark this answers is not in the document]
Comment (aramase) [reply go.mod]:
```
➜ go mod why -m github.com/google/go-cmp
# github.com/google/go-cmp
github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity/v1
github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity/v1.test
github.com/google/go-cmp/cmp
```

File: pkg/nmi/server/server.go [hunk 6]
```diff
@@ -222,41 +217,22 @@ func (s *Server) hostHandler(w http.ResponseWriter, r *http.Request) (ns string)
 		http.Error(w, "parameter resource cannot be empty", http.StatusBadRequest)
 		return
 	}
-	podIDs, identityInCreatedStateFound, err := s.listPodIDsWithRetry(r.Context(), s.KubeClient, podns, podname, rqClientID)
+	podID, err := s.TokenClient.GetIdentities(r.Context(), podns, podname, rqClientID)
```
Comment (kkmsft) [inline pkg/nmi/server/server.go]: Error return from GetIdentities should be detailed enough.

File: test/common/k8s/azureassignedidentity/azureassignedidentity.go [hunk 7]
```diff
@@ -88,10 +88,24 @@ func WaitOnLengthMatched(target int) (bool, error) {
 					return
 				}
 				// len of nil slices is 0, so shouldn't panic here
-				if len(list.Items) == target {
+				if (!checkStatus || target == 0) && len(list.Items) == target {
 					successChannel <- true
 					return
 				}
+				if (checkStatus || target != 0) && list.Items != nil && len(list.Items) == target {
```
Comment (kkmsft) [inline test/common/k8s/azureassignedidentity/azureassignedidentity.go]: can you please add some comments.

Comment (aramase) [reply test/common/k8s/azureassignedidentity/azureassignedidentity.go]: done!

File: pkg/nmi/nmi.go (hunk 3 above)
Comment (aramase) [reply pkg/nmi/nmi.go]: will add readme before release like discussed during code review.

File: pkg/k8s/client.go [hunk 8]
```diff
@@ -121,6 +125,23 @@ func NodeNameFilter(nodeName string) internalinterfaces.TweakListOptionsFunc {
 	}
 }
 
+// GetPod returns pod that matches namespace and name
+func (c *KubeClient) GetPod(namespace, name string) (v1.Pod, error) {
+	// TODO (aramase) wrap this with retries
+	objs := c.PodInformer.GetStore().List()
```
[the remark this answers is not in the document]
Comment (aramase) [reply pkg/k8s/client.go]: switched to using indexer to get by key.

File: pkg/k8s/client.go [hunk 9]
```diff
@@ -121,6 +125,23 @@ func NodeNameFilter(nodeName string) internalinterfaces.TweakListOptionsFunc {
 	}
 }
 
+// GetPod returns pod that matches namespace and name
+func (c *KubeClient) GetPod(namespace, name string) (v1.Pod, error) {
+	// TODO (aramase) wrap this with retries
```
Comment (kkmsft) [inline pkg/k8s/client.go]: Shouldn't we do this before merging since we see this happen in regular operation ?

Comment (aramase) [reply pkg/k8s/client.go]: I don't think this is a blocker for merging the PR. I can add retries in follow up if required.

**Pull request merged**